// staart/lib/hocs/redirect-user.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ooth_client_react_1 = require("ooth-client-react");
var router_1 = require("./router");
var recompose_1 = require("recompose");
exports.default = (function (C) {
    var Redirect = /** @class */ (function (_super) {
        __extends(Redirect, _super);
        function Redirect() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Redirect.prototype.componentDidMount = function () {
            if (this.props.user) {
                this.props.Router.push(this.props.next || '/dashboard');
            }
        };
        Redirect.prototype.componentDidUpdate = function () {
            if (this.props.user) {
                this.props.Router.push(this.props.next || '/dashboard');
            }
        };
        Redirect.prototype.render = function () {
            if (this.props.user) {
                return (React.createElement("div", { style: {
                        maxWidth: '300px',
                        margin: 'auto',
                    } },
                    React.createElement("p", null,
                        "You are logged in. Click ",
                        React.createElement("a", { href: this.props.next || '/dashboard' }, "here"),
                        " if don't get redirected.")));
            }
            return React.createElement(C, __assign({}, this.props));
        };
        return Redirect;
    }(React.Component));
    return recompose_1.compose(ooth_client_react_1.withUser, router_1.default)(Redirect);
});
