// staart/lib/components/forgot-password.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ooth_client_react_1 = require("ooth-client-react");
var recompose_1 = require("recompose");
var i18n_1 = require("../hocs/i18n");
var form_1 = require("./form");
var ForgotPasswordComponent = function (_a) {
    var __ = _a.__;
    return (React.createElement("div", { style: {
            maxWidth: '300px',
            margin: 'auto',
        } },
        React.createElement("h1", null, __('forgot-password.forgot-password')),
        React.createElement(ForgotPasswordForm, null),
        React.createElement("p", null,
            __('forgot-password.back-to'),
            " ",
            React.createElement("a", { href: "/login" }, __('forgot-password.login-page')),
            ".")));
};
var ForgotPassword = recompose_1.compose(i18n_1.default)(ForgotPasswordComponent);
exports.default = ForgotPassword;
var ForgotPasswordFormComponent = /** @class */ (function (_super) {
    __extends(ForgotPasswordFormComponent, _super);
    function ForgotPasswordFormComponent(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {};
        return _this;
    }
    ForgotPasswordFormComponent.prototype.render = function () {
        var _this = this;
        var __ = this.props.__;
        if (this.state.sent) {
            return React.createElement("p", null, __('forgot-password.password-reset-email-sent'));
        }
        return (React.createElement(form_1.default, { onSubmit: function () {
                var username = _this.username.value;
                _this.props.oothClient
                    .method('local', 'forgot-password', {
                    username: username,
                })
                    .then(function (_a) {
                    var message = _a.message;
                    _this.setState({
                        message: message,
                        sent: true,
                        state: 'success',
                    });
                })
                    .catch(function (e) {
                    _this.setState({
                        state: 'error',
                        message: e.message,
                    });
                });
            }, state: this.state.state, message: this.state.message, submitLabel: __('forgot-password.send-reset-password-email') },
            React.createElement("div", { className: "form-group" },
                React.createElement("label", { htmlFor: "username" }, __('forgot-password.username-or-email')),
                React.createElement("input", { type: "username", className: "form-control", id: "username", placeholder: __('forgot-password.username-or-email'), ref: function (username) {
                        _this.username = username;
                    } }))));
    };
    return ForgotPasswordFormComponent;
}(React.Component));
var ForgotPasswordForm = recompose_1.compose(ooth_client_react_1.withOoth, i18n_1.default)(ForgotPasswordFormComponent);
