// staart/lib/components/login-facebook.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var recompose_1 = require("recompose");
var ooth_client_react_1 = require("ooth-client-react");
var i18n_1 = require("../hocs/i18n");
var FacebookComponent = /** @class */ (function (_super) {
    __extends(FacebookComponent, _super);
    function FacebookComponent(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {};
        return _this;
    }
    FacebookComponent.prototype.componentDidMount = function () {
        var _this = this;
        if (typeof window.FB === 'undefined') {
            var $script = require('scriptjs');
            $script('//connect.facebook.net/en_US/sdk.js', function () {
                window.FB.init({
                    appId: _this.props.clientId,
                    version: 'v2.10',
                });
                _this.setState({
                    show: true,
                });
            });
        }
        else {
            this.setState({
                show: true,
            });
        }
    };
    FacebookComponent.prototype.render = function () {
        var _this = this;
        if (!this.state.show) {
            return null;
        }
        return (React.createElement("div", { className: "form-group" },
            React.createElement("button", { onClick: function () {
                    window.FB.login(function (response) {
                        _this.props.oothClient
                            .authenticate('facebook', 'login', {
                            access_token: response.authResponse.accessToken,
                            remember: _this.props.remember,
                        })
                            .catch(function (e) {
                            _this.setState({
                                state: 'error',
                                message: e.message,
                            });
                        });
                    }, {
                        scope: 'email',
                    });
                }, type: "button", className: "btn btn-block btn-default", style: {
                    backgroundColor: '#3B5998',
                    border: '#3B5998',
                    color: 'white',
                } }, this.props.label ? this.props.label : this.props.__('login-facebook.login-with-facebook')),
            this.state.state === 'error' && (React.createElement("div", { className: "alert alert-danger", role: "alert" }, this.state.message))));
    };
    return FacebookComponent;
}(React.Component));
var Facebook = recompose_1.compose(ooth_client_react_1.withOoth, i18n_1.default)(FacebookComponent);
exports.default = Facebook;
