// staart/lib/components/login-local.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var recompose_1 = require("recompose");
var ooth_client_react_1 = require("ooth-client-react");
var i18n_1 = require("../hocs/i18n");
var form_1 = require("./form");
var LoginFormComponent = /** @class */ (function (_super) {
    __extends(LoginFormComponent, _super);
    function LoginFormComponent(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {};
        return _this;
    }
    LoginFormComponent.prototype.render = function () {
        var _this = this;
        var __ = this.props.__;
        return (React.createElement(form_1.default, { onSubmit: function (e) {
                e.preventDefault();
                var username = _this.username.value;
                var password = _this.password.value;
                var remember = _this.props.remember;
                _this.props.oothClient
                    .authenticate('local', 'login', {
                    username: username,
                    password: password,
                    remember: remember,
                })
                    .then(function () {
                    _this.setState({
                        message: __('login-local.logged-in'),
                        state: 'success',
                    });
                })
                    .catch(function (e) {
                    _this.setState({
                        state: 'error',
                        message: e.message,
                    });
                });
            }, state: this.state.state, message: this.state.message, submitLabel: __('login-local.login') },
            React.createElement("div", { className: "form-group" },
                React.createElement("label", { htmlFor: "username" }, __('login-local.username-or-email')),
                React.createElement("input", { type: "string", className: "form-control", id: "username", name: "username", placeholder: "Username", ref: function (username) {
                        _this.username = username;
                    } })),
            React.createElement("div", { className: "form-group" },
                React.createElement("label", { htmlFor: "password" }, __('login-local.password')),
                React.createElement("input", { type: "password", className: "form-control", id: "password", name: "password", placeholder: "******", ref: function (password) {
                        _this.password = password;
                    } }),
                React.createElement("a", { href: "/forgot-password" }, __('login-local.forgot-your-password')))));
    };
    return LoginFormComponent;
}(React.Component));
var LoginForm = recompose_1.compose(ooth_client_react_1.withOoth, i18n_1.default)(LoginFormComponent);
exports.default = LoginForm;
