// staart/lib/components/register-local.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ooth_client_react_1 = require("ooth-client-react");
var recompose_1 = require("recompose");
var i18n_1 = require("../hocs/i18n");
var form_1 = require("./form");
var RegisterFormComponent = /** @class */ (function (_super) {
    __extends(RegisterFormComponent, _super);
    function RegisterFormComponent(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {};
        return _this;
    }
    RegisterFormComponent.prototype.render = function () {
        var _this = this;
        var __ = this.props.__;
        return (React.createElement(form_1.default, { onSubmit: function () {
                var email = _this.email.value;
                var password = _this.password.value;
                _this.props.oothClient
                    .method('local', 'register', {
                    email: email,
                    password: password,
                })
                    .then(function () {
                    return _this.props.oothClient.authenticate('local', 'login', {
                        password: password,
                        username: email,
                    });
                })
                    .then(function () {
                    _this.setState({
                        state: 'success',
                        message: __('login-local.logged-in'),
                    });
                })
                    .catch(function (e) {
                    _this.setState({
                        state: 'error',
                        message: e.message,
                    });
                });
            }, state: this.state.state, message: this.state.message, submitLabel: __('register-local.register') },
            React.createElement("div", { className: "form-group" },
                React.createElement("label", { htmlFor: "email" }, __('register-local.email')),
                React.createElement("input", { type: "email", className: "form-control", id: "email", name: "email", placeholder: "Email", ref: function (email) {
                        _this.email = email;
                    } })),
            React.createElement("div", { className: "form-group" },
                React.createElement("label", { htmlFor: "password" }, __('register-local.password')),
                React.createElement("input", { type: "password", className: "form-control", id: "password", name: "password", placeholder: "******", ref: function (password) {
                        _this.password = password;
                    } }))));
    };
    return RegisterFormComponent;
}(React.Component));
var RegisterForm = recompose_1.compose(ooth_client_react_1.withOoth, i18n_1.default)(RegisterFormComponent);
exports.default = RegisterForm;
