// staart/lib/components/verify-email.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var ooth_client_react_1 = require("ooth-client-react");
var i18n_1 = require("../hocs/i18n");
var recompose_1 = require("recompose");
var VerifyEmailComponent = function (_a) {
    var __ = _a.__, token = _a.token, userId = _a.userId;
    return (React.createElement("div", { style: {
            maxWidth: '300px',
            margin: 'auto',
        } },
        React.createElement("h1", null, __('verify-email.verify-email')),
        React.createElement(VerifyEmailForm, { token: token, userId: userId })));
};
var VerifyEmail = recompose_1.compose(i18n_1.default)(VerifyEmailComponent);
exports.default = VerifyEmail;
var VerifyEmailFormComponent = /** @class */ (function (_super) {
    __extends(VerifyEmailFormComponent, _super);
    function VerifyEmailFormComponent(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {};
        return _this;
    }
    VerifyEmailFormComponent.prototype.componentDidMount = function () {
        var _this = this;
        var __ = this.props.__;
        if (!this.props.token) {
            console.error(__('verify-email.no-token-specified'));
            return;
        }
        if (!this.props.userId) {
            console.error(__('verify-email.no-userid-specified'));
        }
        this.props.oothClient
            .method('local', 'verify', {
            token: this.props.token,
            userId: this.props.userId,
        })
            .then(function (_a) {
            var message = _a.message;
            _this.setState({
                message: message,
                verified: true,
                state: 'success',
            });
        })
            .catch(function (e) {
            _this.setState({
                state: 'error',
                message: e.message,
            });
        });
    };
    VerifyEmailFormComponent.prototype.render = function () {
        var __ = this.props.__;
        if (!this.props.token) {
            return React.createElement("p", null, __('verify-email.no-token-specified'));
        }
        if (!this.props.userId) {
            return React.createElement("p", null, __('verify-email.no-userid-specified'));
        }
        if (this.state.state === 'error') {
            return React.createElement("p", null, this.state.message);
        }
        if (this.state.verified) {
            return React.createElement("p", null, __('verify-email.email-verified'));
        }
        return React.createElement("p", null, __('verify-email.verifying-email'));
    };
    return VerifyEmailFormComponent;
}(React.Component));
var VerifyEmailForm = recompose_1.compose(ooth_client_react_1.withOoth, i18n_1.default)(VerifyEmailFormComponent);
